File: 0041-first-missing-positive/0041-first-missing-positive.py
class Solution:
    def firstMissingPositive(self, nums: List[int]) -> int:
        n  = len(nums)
        if len(nums) == 0:
            return 1
        
        last_pos_idx = self.partition(nums)

        for i in range(last_pos_idx):
            val = abs(nums[i])
            if 1 <= val and val <= last_pos_idx:
                if nums[val - 1] > 0:
                    nums[val - 1] = -nums[val - 1]
        
        for i in range(1, last_pos_idx + 1):
            if nums[i - 1] >= 0:
                return i
        
        return last_pos_idx + 1

    def partition(self, nums):
        # The pivot is known 0
        l = 0
        for r in range(len(nums)):
            if nums[r] > 0:
                self.swap(nums, l , r)
                l += 1
        return l
    def swap(self,nums,start, end):
        temp = nums[start]
        nums[start]  = nums[end]
        nums[end] = temp
        

File: 0041-first-missing-positive/0041-first-missing-positive.java
class Solution {
   public int firstMissingPositive(int[] nums) {
        // O(3n) and O(1) 
        if(nums.length==0) return 1;

        int lastPositiveindex = partition(nums);
        
        for(int i=0;i<lastPositiveindex;i++){
            int k = Math.abs(nums[i]);
            if(1 <= k && k <= lastPositiveindex) {
                if(nums[k - 1] > 0) {
                    nums[k - 1] = - nums[k - 1];
                }
            }
        }
        for(int i= 1;i < lastPositiveindex + 1;i++){
            if(nums[i - 1] >= 0){
                return i;
            }
        }
        return lastPositiveindex + 1;
    }

    private int partition(int[] nums) {
        int l = 0;
        for(int r = 0; r < nums.length; r++){
            if(nums[r] > 0){
                swap(nums, r , l);
                l++;
            }
        }
        return l;
    }
    private void swap(int[] nums, int i, int j) {
        int temp = nums[i];
        nums[i]=nums[j];
        nums[j]=temp;
    }
}